sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,217 @@
+        how the values of the corresponding feature are binned. For a pair
+        of adjacent values in a column, a, b, values between [a, b) will
+        have the same discretization value.
+
+    Example
+    -------
+    >>> X = [[-2, 1, -4,   -1], \
+             [-1, 2, -3, -0.5], \
+             [ 0, 3, -2,  0.5], \
+             [ 1, 4, -1,    2]]
+    >>> dis = KBinsDiscretizer(n_bins=3)
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+    >>> dis.cut_points_ # doctest: +NORMALIZE_WHITESPACE
+    array([[-1., 2., -3., 0.],
+           [ 0., 3., -2., 1.]])
+    >>> dis.transform(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    array([[0, 0, 0, 0],
+           [1, 1, 1, 0],
+           [2, 2, 2, 1],
+           [2, 2, 2, 2]])
+
+    Categorical features can be specified in the constructor.
+
+    >>> dis = KBinsDiscretizer(n_bins=3, categorical_features=[1])
+    >>> dis.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    KBinsDiscretizer(...)
+
+    For each feature mentioned as a categorical feature, this feature
+    will be represented as nan values in the discretizer's cut points.
+    >>> dis.cut_points_ # doctest: +NORMALIZE_WHITESPACE
+    array([[ -1., nan, -3., 0.],
+           [  0., nan, -2., 1.]])
+
+    The discretization process will not touch any column marked as categorical.
+    >>> dis.transform(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    array([[ 0., 1., 0., 0.],
+           [ 1., 2., 1., 0.],
+           [ 2., 3., 2., 1.],
+           [ 2., 4., 2., 2.]])
+    """
+
+    def __init__(self, n_bins=2, categorical_features=None):
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+        # Attributes
+        self.mins_ = None
+        self.maxes_ = None
+        self.n_features_ = None
+        self.cut_points_ = None
+        self.categorical_features_set_ = None
+
+    def fit(self, X, y=None):
+        """Fits the estimator.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            The input data containing features to be discretized. Indexes
+            of categorical features are indicated by categorical_features
+            from the constructor.
+
+        Returns
+        -------
+        self
+        """
+
+        if self.n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins. Received {0}, expected at least 2."
+                             .format(self.n_bins))
+
+        X = check_array(X, dtype=float, ensure_2d=False)
+
+        if X.ndim == 1:
+            raise ValueError("Input array must be 2D shape.")
+
+        self._check_categorical_features(self.categorical_features, X.shape[1])
+
+        self._fit(X)
+        return self
+
+    @staticmethod
+    def _check_categorical_features(cat_features, n_features):
+        """Only allow integer indices, or None.
```

Comment:
This description is confusing because "integer indices" could be referring to the feature values!